### PR TITLE
CUDNN v8 support

### PR DESCRIPTION
**Pull request by nluehr**
Enables building against the CUDNN v8 library.

Attn: @reedwm 

Note this PR can be applied separately from the cuda 11 PR (39576), but both modify find_cuda_config.py. As a result, they will conflcit on find_cuda_config.py.gz.base64 and that file will need to be regenerated via compress_find_cuda_config.py when the second PR is merged.

### CODE REVIEW

File: third_party/gpus/cuda_configure.bzl [hunk 1]
```diff
@@ -967,11 +967,30 @@ def _create_local_cuda_repository(repository_ctx):
...
+    if [int(x) for x in cuda_config.cudnn_version.split(".")] < [8, 0]:
+      cudnn_headers = ["cudnn.h"]
+    else:
+      cudnn_headers = ["cudnn_adv_infer.h",
+                       "cudnn_adv_train.h",
+                       "cudnn_cnn_infer.h",
+                       "cudnn_cnn_train.h",
+                       "cudnn_ops_infer.h",
+                       "cudnn_ops_train.h",
+                       "cudnn.h",
```
Comment (freedomtan) [inline third_party/gpus/cuda_configure.bzl:979]: there are trailing spaces after `"cudnn.h."`

**Comment by chsigg** [comment]
Hi Nathan, thank you very much for this PR. I'm going to temporarily mark it as 'ready to pull'. This will simplify testing it on our side and merging it.

**Review by pkanwar23 — APPROVED** [review]

**Comment by rthadur** [comment]
@nluehr can you please resolve conflicts ?

**Comment by chsigg** [comment]
> @nluehr can you please resolve conflicts ?

Hi Nathan, you don't need to resolve, it's easier to merge this manually from our side and I'm actively working on that. The CUDA 11 changes have been merged already, and hopefully this one will land until Monday as well. There was some trouble with compiler commands going over the Windows character limit and unrelated breakages.

I'm removing the 'ready to pull' tag again so the automated tools don't try to submit it as well.

**Comment by YashasSamaga** [comment]
What was the reason for not using `WINOGRAD_NONFUSED` algorithms?

**Comment by YashasSamaga** [comment]
> What was the reason for not using WINOGRAD_NONFUSED algorithms?

Just leaving a comment here so that future visitors can get the answer to the above quesiton.

https://forums.developer.nvidia.com/t/variations-in-heuristics-for-selecting-conv-algorithm-in-different-apis/139730/2

**Comment by nluehr** [comment]
The reason was just to match the behavior with cudnn v7.

**Pull request merged**